ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java

```diff
@@ -417,6 +440,291 @@ Map<String, DataSource> getDataSources() {
+    // 1. Attempt to insert into the table. This is attempted first since it is the most common case.
+    try (PreparedStatement statement = connection.prepareStatement(INSERT_QUERY)) {
+      statement.setInt(1, accountId);
+      statement.setInt(2, containerId);
+      statement.setString(3, record.getBlobName());
+      statement.setBytes(4, Base64.decodeBase64(record.getBlobId()));
+      if (record.getExpirationTimeMs() != Utils.Infinite_Time) {
+        statement.setTimestamp(5, new Timestamp(record.getExpirationTimeMs()));
+      } else {
+        statement.setTimestamp(5, null);
+      }
+      statement.executeUpdate();
+    } catch (SQLIntegrityConstraintViolationException e) {
+      rowAlreadyExists = true;
+    }
+    // 2. If the row already exists, attempt an update, checking that the update should be allowed (the current
+    //    row in the db represents a blob that was soft deleted or expired. Since soft deleted records are not
+    //    deleted from the table before a long retention period, there is no risk of the row not existing when
+    //    this query is run.
+    if (rowAlreadyExists) {
+      try (PreparedStatement statement = connection.prepareStatement(UPDATE_IF_DELETED_OR_EXPIRED_QUERY)) {
+        // fields to update
+        statement.setBytes(1, Base64.decodeBase64(record.getBlobId()));
+        if (record.getExpirationTimeMs() != Utils.Infinite_Time) {
+          statement.setTimestamp(2, new Timestamp(record.getExpirationTimeMs()));
+        } else {
+          statement.setTimestamp(2, null);
+        }
+        // primary key fields
+        statement.setInt(3, accountId);
+        statement.setInt(4, containerId);
+        statement.setString(5, record.getBlobName());
+        // the number of rows found will be 0 if there is currently an entry for this blob name that is
+        // not expired or deleted.
+        if (statement.executeUpdate() == 0) {
+          throw buildException("PUT: Blob still alive", RestServiceErrorCode.Conflict, record.getAccountName(),
+              record.getContainerName(), record.getBlobName());
+        }
+      }
+    }
+    return new PutResult(record);
+  }
+
+  private PutResult run_put_v2(NamedBlobRecord record, short accountId, short containerId, Connection connection)
+      throws Exception {
+    // 1. Attempt to insert into the table. This is attempted first since it is the most common case.
+    try (PreparedStatement statement = connection.prepareStatement(INSERT_QUERY_V2)) {
+      statement.setInt(1, accountId);
+      statement.setInt(2, containerId);
+      statement.setString(3, record.getBlobName());
+      statement.setBytes(4, Base64.decodeBase64(record.getBlobId()));
+      if (record.getExpirationTimeMs() != Utils.Infinite_Time) {
+        statement.setTimestamp(5, new Timestamp(record.getExpirationTimeMs()));
+      } else {
+        statement.setTimestamp(5, null);
+      }
+      statement.setLong(6, buildVersion());
```

Comment:
looks like run_put_v2 will take all the put request, but for run_put, it will return conflict when blob still alive.
So when you do get, would it be different? So you will never know if db and db_v2 is consistent or not?

Reply:
Right, that's for supporting versioned named-blob, so run_put_v2, will create new version instead of raise conflict. 
But look at the put method, I don't catch any exception from 'run_put', so if it failed in 'run_put' due to conflicts, then it won't call run_put_v2. Therefore, it should be ok regarding the db vs db_v2 comparison.
```
          putResult = run_put(record, accountId, containerId, connection);
          try {
            run_put_v2(record, accountId, containerId, connection);
          } catch (Exception e) {
            logger.error("NAMED PUT: Data Error: old table succeed while new table failed: ", e);
            metricsRecoder.namedDataErrorPutCount.inc();
          }
```

Reply:
Yeah, I understand it will be good for PUT. But for GET, let's say when people trying to get a named blob, but the record is different since the old DB won't have the latest value. Would that be a problem? I thought finally you will compare all the record and make sure data in DB is consistent with data in DB_V2?

Reply:
As we chatted, the run_put_v2 won't run if the run_put failed. So we are good with this.